### ISSUE TITLE
Better benchmarks

### ISSUE THREAD

**Issue by bvisness** [HandmadeMath/Handmade-Math#78]
We should come up with a proper benchmarking solution for Handmade Math. Since we care a lot about performance, it would be worth having automatic benchmarks that run on every build.

It would be nice if our benchmarking tools could display historical data to us in some way so that we can easily judge the speedup of a particular change.

This issue inspired by @IJzerbaard's very helpful quick-bench benchmarks.

**Comment by StrangeZak** [comment]
I agree are there any web tools we can just send some data to and it will graph it over time.

I already have a landing page for handmade math: http://handmade-math.com/

We could extend this to include benchmark data for us since its just on a normal digitalocean droplet linux server. Not sure if there are any 3rd party tools that automatically do this already, or if its something we'd have to write. Either way sounds like a great idea. Probably could be reusable for other things also

**Comment by RandyGaul** [comment]
Personally I would love to see a benchmark by taking a chunk of code out of an actual game or product. Maybe like a particle system, or a physics solver, or maybe like an audio mixer. Anything would work. Plotting timings over time for a real-world use-case could be very interesting and educational.

Also besides just benchmarks, having disassembly of a function can make a good data point. For example here Mitton has some nice disassembly of a raycasting function: http://www.codersnotes.com/notes/maths-lib-2016/

The disassembly is nice to identify stack bleeding (when not enough regisers are available and excessive push/pop to/from stack can occur). This kind of thing may be much simpler than running a full time-perf test, since it's simply compiling a function and inspecting the output. Disassembly can be posted into a readme markup file along with the original function definition, and create a very educational resource.

Just some thoughts. I like this library and want to see it grow :)

I could be willing to donate and port some physics solver code or something similar, if it would help.

**Comment by StrangeZak** [comment]
Hi @RandyGaul 

I completely agree with everything you've said. Real world bench marking is way better than just making a big test that multiplies matrices 1 million times for example.

I feel like there's a few different types of info we'd want out of a benchmark tool. 

1. The ability to see optimized and unoptimized disassembly  on a per function basis for a few separate compilers (GCC and Clang?)  
2. The amount of cycles a operation takes (i.e: M *M)
3. Data storing to see how data is trending over time
4.  Historic Graphs (Who doesn't love graphs 😛 )
5.  How the code performs in a _real_ application against other libraries 

And as you said this _real_ application could be anything like a real time renderer, physics simulation, audio mixer, anything that uses a lot of Vectors or Matrices ideally.
